nucleus/common/internal-api/src/main/java/org/glassfish/internal/api/Globals.java

```diff
@@ -144,7 +154,50 @@ public static ServiceLocator getStaticHabitat() {
 
         return defaultHabitat;
     }
+
+    /**
+     * Returns the static service locator, but using the classloader used to load
+     * the modules/ directory. This is useful when you need access to classes from
+     * the modules/ directory from client code.
+     * 
+     * @return the static service locator with modules loaded
+     * @throws MalformedURLException if the URL to a module is invalid
+     */
+    public static ServiceLocator getStaticHabitatWithModules() throws MalformedURLException {
+        if (defaultHabitatWithModules == null) {
+            synchronized(STATIC_LOCK) {
+                if (defaultHabitatWithModules == null) {
+                    defaultHabitatWithModules = createStaticHabitatWithModules();
+                }
+            }
+        }
+        return defaultHabitatWithModules;
+    }
     
+    private static ServiceLocator createStaticHabitatWithModules() throws MalformedURLException {
+        // get the list of JAR files from the modules directory
+        ArrayList<URL> urls = new ArrayList<>();
+        File idir = new File(System.getProperty(SystemPropertyConstants.INSTALL_ROOT_PROPERTY));
+        File mdir = new File(idir, "modules");
+        for (File f : mdir.listFiles()) {
+            if (f.toString().endsWith(".jar")) {
+                urls.add(f.toURI().toURL());
+            }
+        }
+
+        final URL[] urlsA = urls.toArray(new URL[urls.size()]);
+
+        ClassLoader cl = (ClassLoader) AccessController.doPrivileged(new PrivilegedAction<Object>() {
+            @Override
+            public Object run() {
+                return new URLClassLoader(urlsA, Globals.class.getClassLoader());
+            }
+        });
+
+        ModulesRegistry registry = new StaticModulesRegistry(cl);
+        return registry.createServiceLocator("default");
```

Comment:
I doubt that this is correct - the StaticModulesRegistry will exist only until it would be destroyed by the GC? How long after exiting this method will locator work correctly?

Reply:
What do you mean by this?

Reply:
StaticModuleRegistry serves as a register of modules, it has also shutdown method ... it does not implement Closeable, because the implementation is very old, but I tried to investigate a bit ... it has shutdown() method and contains a Register which can use timers which are interrupted on shutdown ... so at least it should call shutdown in finally. But then I'm not sure if the provided service object is still valid or not (it might be also in inconsistent state if it would have a reference to the registry and trying to use it).
The registry does not have finalize method implementation, so when it is garbage collected, nothing stops the timer ... so this can lead to resource leak ... (how do we work with *ModuleRegistry in other cases?)

Reply:
Hmm I hadn't spotted the shutdown method. Do you think a static reference should not be kept then?

Reply:
Why it cannot just use Globals.getStaticBaseServiceLocator()?

Reply:
Because then the xml tags aren't recognised. The method is javadocced, I just put it somewhere common so that both sources could use the same function. If it's a big issue, I can just duplicate the code

Reply:
Hmm, tags are not recognized with getStaticBaseServiceLocator? That's weird ... because the method should do the same ... or I am missing something. I will try to checkout this PR after I will get rid of P4 build ...